File: java/ods/FastArrayStack.java
package ods;

/**
 * This subclass of ArrayStack overrides some of its functions to make
 * them faster. In particular, it uses System.arraycopy() rather than for
 * loops copy elements between arrays and to shift them within arrays.
 * @author morin
 *
 * @param <T>
 */
public class FastArrayStack<T> extends ArrayStack<T> {
	public FastArrayStack(Class<T> t0) {
		super(t0);
	}

	protected void resize() {
		T[] b = f.newArray(Math.max(2*n,1));
		System.arraycopy(a, 0, b, 0, n);
		a = b;
	}

	protected void resize(int nn) {
		T[] b = f.newArray(nn);
		System.arraycopy(a, 0, b, 0, n);
		a = b;
	}

	public void add(int i, T x) {
		if (i < 0 || i > n) throw new IndexOutOfBoundsException();
		if (n + 1 > a.length) resize();
		System.arraycopy(a, i, a, i+1, n-i); 
		a[i] = x;
		n++;
	}
	
	public T remove(int i) {
		if (i < 0 || i > n - 1) throw new IndexOutOfBoundsException();
		T x = a[i];
		System.arraycopy(a, i+1, a, i, n-i-1);
		n--; 
		if (a.length >= 3*n) resize();
		return x;
	}
}

File: cpp/FastArrayStack.cpp
/*
 * FastArrayStack.cpp
 *
 *  Created on: 2011-11-23
 *      Author: morin
 */
#include <string.h>
#include "FastArrayStack.h"
#include "utils.h"

namespace ods {

template<class T>
FastArrayStack<T>::FastArrayStack() : ArrayStack<T>() {
}

template<class T>
FastArrayStack<T>::~FastArrayStack() {
}

template<class T>
void FastArrayStack<T>::resize()
{
	length = max(1, 2*n);
	T *b = new T[length];
	memcpy(b, a, n*sizeof(T));
	delete a;
	a = b;
}

template<class T>
void FastArrayStack<T>::add(int i, T x) {
	if (n + 1 > length) resize();
	memcpy(&a[i+1], &a[i], (n-i)*sizeof(T));
	a[i] = x;
	n++;
}

template<class T>
T FastArrayStack<T>::remove(int i)
{
    T x = a[i];
	memcpy(&a[i], &a[i+1], (n-i-1)*sizeof(T));
	n--;
	if (length >= 3 * n) resize();
	return x;
}

template FastArrayStack<int>::FastArrayStack();
template FastArrayStack<int>::~FastArrayStack();
template void FastArrayStack<int>::add(int,int);
template int FastArrayStack<int>::remove(int);



} /* namespace ods */
